File: conformance/cdc-profile/v1/scalar_fastcdc.py
"""Independent scalar FastCDC oracle owned by the v1 conformance corpus."""

from __future__ import annotations

from collections.abc import Iterable, Iterator
import itertools

MINIMUM, TARGET, MAXIMUM = 16_384, 65_536, 262_144
NORMALIZATION, STATE_WIDTH, SEED = 2, 64, 0
SHORT_MASK = 0x0000_D907_0753_7000
LONG_MASK = 0x0000_D903_1353_0000
U64_MASK = (1 << 64) - 1


def reference_boundaries(source: bytes, gear: tuple[int, ...]) -> tuple[int, ...]:
    """Return absolute exclusive ends from the one-shot scalar algorithm."""
    ends: list[int] = []
    start = 0
    while start < len(source):
        remaining = len(source) - start
        if remaining <= MINIMUM:
            start += remaining
            ends.append(start)
            continue
        limit = min(remaining, MAXIMUM)
        center = min(TARGET, limit)
        fingerprint = SEED
        cut = limit
        for position in range(MINIMUM, center):
            fingerprint = ((fingerprint << 1) + gear[source[start + position]]) & U64_MASK
            if fingerprint & SHORT_MASK == 0:
                cut = position
                break
        else:
            for position in range(center, limit):
                fingerprint = ((fingerprint << 1) + gear[source[start + position]]) & U64_MASK
                if fingerprint & LONG_MASK == 0:
                    cut = position
                    break
        start += cut
        ends.append(start)
    return tuple(ends)


class StreamingChunker:
    """Scalar oracle whose feed partitioning cannot affect its state machine."""

    def __init__(self, gear: tuple[int, ...]) -> None:
        self.gear = gear
        self.current = bytearray()
        self.completed: list[bytes] = []
        self.fingerprint = SEED

    def feed(self, part: bytes) -> None:
        """Advance state without treating an empty part as EOF."""
        for value in part:
            self._feed_byte(value)

    def _feed_byte(self, value: int) -> None:
        position = len(self.current)
        if position < MINIMUM:
            self.current.append(value)
            return
        self.fingerprint = ((self.fingerprint << 1) + self.gear[value]) & U64_MASK
        mask = SHORT_MASK if position < TARGET else LONG_MASK
        if self.fingerprint & mask == 0:
            self._emit()
            self.current.append(value)
            return
        self.current.append(value)
        if len(self.current) == MAXIMUM:
            self._emit()

    def _emit(self) -> None:
        if not self.current:
            raise ValueError("streaming oracle cannot emit an empty chunk")
        self.completed.append(bytes(self.current))
        self.current.clear()
        self.fingerprint = SEED

    def finish(self) -> tuple[list[bytes], tuple[int, ...]]:
        """Declare EOF and return chunks plus their absolute exclusive ends."""
        if self.current:
            self._emit()
        total = 0
        ends = []
        for chunk in self.completed:
            total += len(chunk)
            ends.append(total)
        return self.completed, tuple(ends)


def scheduled_parts(source: bytes, sizes: Iterable[int]) -> Iterator[bytes]:
    """Partition source bytes according to a positive repeating size schedule."""
    offset = 0
    iterator = iter(sizes)
    while offset < len(source):
        size = next(iterator)
        if size <= 0:
            raise ValueError("partition sizes must be positive")
        end = min(len(source), offset + size)
        yield source[offset:end]
        offset = end


def boundary_adjacent_parts(source: bytes, ends: tuple[int, ...]) -> list[bytes]:
    """Partition immediately before, at, and after every expected boundary."""
    points = {0, len(source)}
    for boundary in ends:
        points.update({max(0, boundary - 1), boundary, min(len(source), boundary + 1)})
    ordered = sorted(points)
    return [source[left:right] for left, right in itertools.pairwise(ordered) if left < right]


def probe_fingerprint(source: bytes, position: int, gear: tuple[int, ...]) -> int:
    """Return the reset fingerprint after including probe byte at position."""
    fingerprint = SEED
    for offset in range(MINIMUM, position + 1):
        fingerprint = ((fingerprint << 1) + gear[source[offset]]) & U64_MASK
    return fingerprint

File: xtask/src/protocol_conformance/cdc_profile/scalar_fastcdc.rs
//! This module owns the independent scalar and streaming `FastCDC` oracle.

use std::mem;

use super::{GearTable, LONG_MASK, MAXIMUM, MINIMUM, SEED, SHORT_MASK, TARGET};
use crate::protocol_conformance::ConformanceError;

pub(super) fn reference_boundaries(
    source: &[u8],
    gear: &GearTable,
) -> Result<Vec<usize>, ConformanceError> {
    let mut ends = Vec::new();
    let mut start = 0_usize;
    while start < source.len() {
        let remaining = source
            .len()
            .checked_sub(start)
            .ok_or_else(|| ConformanceError::violation("CDC remaining length underflow"))?;
        if remaining <= MINIMUM {
            start = start
                .checked_add(remaining)
                .ok_or_else(|| ConformanceError::violation("CDC final boundary overflow"))?;
            ends.push(start);
            continue;
        }
        let limit = remaining.min(MAXIMUM);
        let center = TARGET.min(limit);
        let cut = scan_cut(source, gear, start, center, limit)?;
        start = start
            .checked_add(cut)
            .ok_or_else(|| ConformanceError::violation("CDC boundary overflow"))?;
        ends.push(start);
    }
    Ok(ends)
}

fn scan_cut(
    source: &[u8],
    gear: &GearTable,
    start: usize,
    center: usize,
    limit: usize,
) -> Result<usize, ConformanceError> {
    let mut fingerprint = SEED;
    for position in MINIMUM..center {
        fingerprint = update(fingerprint, source_byte(source, start, position)?, gear)?;
        if fingerprint & SHORT_MASK == 0 {
            return Ok(position);
        }
    }
    for position in center..limit {
        fingerprint = update(fingerprint, source_byte(source, start, position)?, gear)?;
        if fingerprint & LONG_MASK == 0 {
            return Ok(position);
        }
    }
    Ok(limit)
}

fn source_byte(source: &[u8], start: usize, position: usize) -> Result<u8, ConformanceError> {
    let index = start
        .checked_add(position)
        .ok_or_else(|| ConformanceError::violation("CDC probe offset overflow"))?;
    source
        .get(index)
        .copied()
        .ok_or_else(|| ConformanceError::violation("CDC probe is outside its source"))
}

fn update(fingerprint: u64, byte: u8, gear: &GearTable) -> Result<u64, ConformanceError> {
    let value = gear
        .get(usize::from(byte))
        .copied()
        .ok_or_else(|| ConformanceError::violation("Gear byte index is absent"))?;
    Ok(fingerprint.wrapping_shl(1).wrapping_add(value))
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub(super) struct StreamingSnapshot {
    current: Vec<u8>,
    completed: Vec<Vec<u8>>,
    fingerprint: u64,
}

pub(super) struct StreamingChunker<'a> {
    gear: &'a GearTable,
    current: Vec<u8>,
    completed: Vec<Vec<u8>>,
    fingerprint: u64,
}

impl<'a> StreamingChunker<'a> {
    pub(super) const fn new(gear: &'a GearTable) -> Self {
        Self {
            gear,
            current: Vec::new(),
            completed: Vec::new(),
            fingerprint: SEED,
        }
    }

    pub(super) fn feed(&mut self, part: &[u8]) -> Result<(), ConformanceError> {
        for value in part {
            self.feed_byte(*value)?;
        }
        Ok(())
    }

    fn feed_byte(&mut self, value: u8) -> Result<(), ConformanceError> {
        let position = self.current.len();
        if position < MINIMUM {
            self.current.push(value);
            return Ok(());
        }
        self.fingerprint = update(self.fingerprint, value, self.gear)?;
        let mask = if position < TARGET {
            SHORT_MASK
        } else {
            LONG_MASK
        };
        if self.fingerprint & mask == 0 {
            self.emit()?;
            self.current.push(value);
            return Ok(());
        }
        self.current.push(value);
        if self.current.len() == MAXIMUM {
            self.emit()?;
        }
        Ok(())
    }

    fn emit(&mut self) -> Result<(), ConformanceError> {
        if self.current.is_empty() {
            return Err(ConformanceError::violation(
                "streaming oracle cannot emit an empty chunk",
            ));
        }
        self.completed.push(mem::take(&mut self.current));
        self.fingerprint = SEED;
        Ok(())
    }

    /// Flushes a non-empty trailing chunk and finalizes result access.
    ///
    /// Repeated calls are idempotent while no additional bytes are fed.
    pub(super) fn finish(&mut self) -> Result<(), ConformanceError> {
        if !self.current.is_empty() {
            self.emit()?;
        }
        Ok(())
    }

    /// Returns cumulative end offsets for the finalized chunks.
    ///
    /// Call [`Self::finish`] first. A non-empty unfinished tail is refused, and
    /// cumulative-length overflow returns [`ConformanceError`]. The returned
    /// vector allocates storage proportional to the completed chunk count.
    pub(super) fn boundaries(&self) -> Result<Vec<usize>, ConformanceError> {
        self.require_finished()?;
        let mut total = 0_usize;
        self.completed
            .iter()
            .map(|chunk| {
                total = total
                    .checked_add(chunk.len())
                    .ok_or_else(|| ConformanceError::violation("streaming boundary overflow"))?;
                Ok(total)
            })
            .collect()
    }

    /// Materializes the exact finalized byte stream.
    ///
    /// Call [`Self::finish`] first. A non-empty unfinished tail is refused.
    /// Success allocates the complete reconstructed source.
    pub(super) fn reconstruct(&self) -> Result<Vec<u8>, ConformanceError> {
        self.require_finished()?;
        Ok(self.completed.concat())
    }

    fn require_finished(&self) -> Result<(), ConformanceError> {
        if self.current.is_empty() {
            Ok(())
        } else {
            Err(ConformanceError::violation(
                "streaming results require a finished stream",
            ))
        }
    }

    pub(super) fn snapshot(&self) -> StreamingSnapshot {
        StreamingSnapshot {
            current: self.current.clone(),
            completed: self.completed.clone(),
            fingerprint: self.fingerprint,
        }
    }
}

pub(super) fn probe_fingerprint(
    source: &[u8],
    position: usize,
    gear: &GearTable,
) -> Result<u64, ConformanceError> {
    let mut fingerprint = SEED;
    let end = position
        .checked_add(1)
        .ok_or_else(|| ConformanceError::violation("probe position overflow"))?;
    for offset in MINIMUM..end {
        let value = source
            .get(offset)
            .copied()
            .ok_or_else(|| ConformanceError::violation("probe position is outside its source"))?;
        fingerprint = update(fingerprint, value, gear)?;
    }
    Ok(fingerprint)
}

#[cfg(test)]
mod tests {
    use super::{ConformanceError, GearTable, MINIMUM, StreamingChunker, reference_boundaries};

    #[test]
    fn matching_probe_byte_starts_the_next_streaming_chunk() {
        let gear: GearTable = [0; 256];
        let source = vec![0; MINIMUM + 2];
        assert!(matches!(
            reference_boundaries(&source, &gear),
            Ok(ref boundaries) if boundaries == &[MINIMUM, MINIMUM + 2]
        ));
        let mut chunker = StreamingChunker::new(&gear);
        let before = chunker.snapshot();
        assert!(chunker.feed(&[]).is_ok());
        assert_eq!(chunker.snapshot(), before);
        assert!(chunker.feed(&source).is_ok());
        assert!(chunker.finish().is_ok());
        assert!(matches!(
            chunker.boundaries(),
            Ok(ref boundaries) if boundaries == &[MINIMUM, MINIMUM + 2]
        ));
        assert!(matches!(
            chunker.reconstruct(),
            Ok(ref reconstructed) if reconstructed == &source
        ));
    }

    #[test]
    fn unfinished_nonempty_stream_refuses_result_access() {
        let gear: GearTable = [0; 256];
        let mut chunker = StreamingChunker::new(&gear);
        assert!(chunker.feed(b"unfinished").is_ok());
        assert!(matches!(
            chunker.boundaries(),
            Err(ConformanceError::Violation(ref message))
                if message == "streaming results require a finished stream"
        ));
        assert!(matches!(
            chunker.reconstruct(),
            Err(ConformanceError::Violation(ref message))
                if message == "streaming results require a finished stream"
        ));
    }
}
